python-frontend/src/main/java/org/sonar/python/types/RuntimeType.java

```diff
@@ -71,11 +71,11 @@ public boolean canBeOrExtend(String typeName) {
 
   @Override
   public boolean isCompatibleWith(InferredType other) {
-    if (other instanceof RuntimeType) {
-      RuntimeType otherRuntimeType = (RuntimeType) other;
-      String otherFullyQualifiedName = otherRuntimeType.typeClass.fullyQualifiedName();
-      boolean isDuckTypeCompatible = !otherRuntimeType.typeClass.declaredMembers().isEmpty() &&
-        otherRuntimeType.typeClass.declaredMembers().stream().allMatch(m -> typeClass.resolveMember(m.name()).isPresent());
+    if (other instanceof RuntimeType || other instanceof DeclaredType) {
+      ClassSymbol otherTypeClass = InferredTypes.typeSymbols(other).iterator().next();
+      String otherFullyQualifiedName = otherTypeClass.fullyQualifiedName();
+      boolean isDuckTypeCompatible = !otherTypeClass.declaredMembers().isEmpty() &&
```

Comment:
cf. my other comment, I am not sure about the validity of `!otherTypeClass.declaredMembers().isEmpty()` here (although this is not new). From what I can see it's to avoid considering `NoneType` as compatible with other types, but I guess we could do it differently.